docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+```
+
+The above rule looks menacing. Here is an example of matching a record
+in the above example with the variable `log`:
+
+```tla
+IsDefined(eventAOrB) ==
+  LET ElseB(onlyB) ==
+    MatchTag(onlyB, "EventB", LAMBDA b: b.src /= "", LAMBDA elseValue: FALSE)
+  IN  
+  MatchTag(event, "EventA", LAMBDA a: a.val /= -1, ElseB)
+```
+
+The operator `ElseB` looks redundant, as we know that `onlyB` is a singleton
+variant. To this end, we introduce the operator `MatchOnly`:
+
+```tla
+MatchOnly(variant, ThenOper(_)) ==
+  \* fallback untyped implementation
+  ThenOper(variant)
+```
+
+For completeness, we give the inference rule for `MatchOnly`:
+
+```
+variant: Variant(RowCons("<TAG>", r, RowNil))
+ThenOper: r => z
+------------------------------------------------ [variant_match_only]
+MatchOnly(variant, ThenOper):
+  (Variant(RowCons("<TAG>", r, RowNil)),
+   r => z
+  ) => z
+```
+
+It looks like the solution with `MatchTag` and `MatchOnly` introduce a lot of
+boilerplate. However, this is probably the best solution that we can have,
+unless we can extend the grammar of TLA+.
+
+### 5.3. Changes in the model checker
+
+Having precise types for variants, we have two options:
+
+ 1. Keep the current encoding, that is, a variant is encoded as a super-record
+ that contains all possible fields of the member records. The type checker
+ will guarantee that we do not access the fields of the super-record that are
+ not present in the actual record type.
+
+ 1. Implement the
+ [suggestion by Shon Feder](https://github.com/informalsystems/apalache/discussions/789#discussioncomment-1592118).
+
+Although Option 2 looks nicer, we prefer keeping Option 1. The reason is that
+the current implementation introduces the minimal number of constraints
+by mashing all possible fields into a super-record. The alternative solution
+(option 2) would introduce additional constraints, when the spec requires us
+to extract an element from a set.
+
+Recall the example with a set of messages:
+
+```tla
+VARIABLES
+  (*
+    @typeAlias: MESSAGE =
+       [ tag: "1a", bal: Int ]
+     | [ tag: "2a", bal: Int, val: Int ]);
+    @type: Set(MESSAGE);
+   *)  
+  msgs
+```  
+
+Consider an existential quantifier over the variable `msgs`:
+
+```tla
+Next ==
+  \E m \in msgs:
+    P
+```
+
+In the current encoding, `m` is a super-record that contains three fields:
+`tag`, `bal`, and `val`, even if some of these fields are not required by the
+actual type of `m`. In the alternative encoding, `m` is a tuple `tup`,
+which equals to one of the following tuples, depending on the value of the
+field `m.tag`:
+
+```
+tup = IF m.tag = "1a"
+      THEN << { [ tag |-> "1a", bal |-> b ] }, {} >>
+      ELSE << {}, [ tag |-> "2a", bal |-> b, val |-> v ] }, {} >>
+```
+
+Since it is impossible to statically compute the actual type of `m`, the
+rewriting rules would have to replicate the structure of both elements of the
+tuple `tup`. This would lead to a blow-up in the number of constraints.
```

Comment:
Could you show the equivalent here in your proposed encoding? I am having trouble understanding what that's meant to look like.

Reply:
The arena encoding does not change but stays the same. A record will contain all possible fields. In this case, it will be three fields: `tag`, `bal`, and `val`. Since it is a single record, not a tuple of records, the rewriting rules would have to pick values only from that record instead of introducing copies of all possible records in the tuple.